### PR TITLE
Handle errors in autolinking scripts

### PR DESCRIPTION
**Pull request by eduardopelitti**
This PR aims to provide some basic error handling logic to the autolinking scripts by identifying potential drawbacks and managing the execution flow via try/catch statements and the use of throw under specific conditions.

**Example of a successful link:** (mostly unchanged, added info of some steps which were silently completing the tasks)

Screen Shot 2020-06-01 at 17 13 16

**Example of a partial link (either the app is already linked or part of it is):**

Screen Shot 2020-06-01 at 17 14 07

**Examples of failed links (some of the steps fail):**

Screen Shot 2020-06-01 at 18 10 25

Screen Shot 2020-06-01 at 17 36 02

Screen Shot 2020-06-01 at 17 33 42

### CODE REVIEW

File: autolink/postlink/activityLinker.js [hunk 1]
```diff
@@ -1,23 +1,52 @@
...
+var path = require("./path");
 var fs = require("fs");
-var { warnn, logn, infon, debugn } = require("./log");
+var { errorn, warnn, logn, infon, debugn } = require("./log");
 
 class ActivityLinker {
   constructor() {
     this.activityPath = path.mainActivityJava;
+    this.extendNavigationActivity = false;
+    this.removeGetMainComponentName = false;
```
Comment (eduardopelitti) [inline autolink/postlink/activityLinker.js]: I added these status variables to keep tabs on some of the necessary actions that could fail during the process

File: autolink/postlink/activityLinker.js [hunk 2]
```diff
@@ -1,23 +1,52 @@
...
   }
 
   link() {
+    if (!this.activityPath) {
```
Comment (eduardopelitti) [inline autolink/postlink/activityLinker.js:14]: This is one of the most basic checks, checking for the file that will be edited to exist in the given directory.

File: autolink/postlink/activityLinker.js [hunk 3]
```diff
@@ -1,23 +1,52 @@
...
+      return;
+    }
+
     logn("Linking MainActivity...");
-    if (this.activityPath) {
-      var activityContent = fs.readFileSync(this.activityPath, "utf8");
+
+    var activityContent = fs.readFileSync(this.activityPath, "utf8");
+
+    try {
```
Comment (eduardopelitti) [inline autolink/postlink/activityLinker.js:25]: Each of the methods was added in a try/catch statement, errors are thrown from inside these methods.

File: autolink/postlink/activityLinker.js [hunk 4]
```diff
@@ -1,23 +1,52 @@
...
       infon("MainActivity linked successfully!\n");
+    } else if (!this.extendNavigationActivity && !this.removeGetMainComponentName) {
+      errorn(
+        "MainActivity was not linked. Please check the logs above for more information and proceed with manual linking of the MainActivity file in Android:\nhttps://wix.github.io/react-native-navigation/docs/installing#2-update-mainactivityjava"
+      );
     } else {
-      warnn("   MainActivity not found!");
+      warnn(
+        "MainActivity was only partially linked. Please check the logs above for more information and proceed with manual linking for the failed steps:\nhttps://wix.github.io/react-native-navigation/docs/installing#2-update-mainactivityjava"
+      );
```
Comment (eduardopelitti) [inline autolink/postlink/activityLinker.js:49]: These are the messages displayed as the final outcome of each step in the linking process, considering the outcome of running each of the methods above, we get a successful, partially successful or failed step.

File: autolink/postlink/activityLinker.js [hunk 5]
```diff
@@ -27,24 +56,32 @@ class ActivityLinker {
       debugn("   Removing getMainComponentName function");
       return contents.replace(/\/\*\*\s*\n([^\*]|(\*(?!\/)))*\*\/\s*@Override\s*protected\s*String\s*getMainComponentName\s*\(\)\s*{\s*return.+\s*\}/, "");
     }
+    warnn("   getMainComponentName function was not found.");
```
Comment (eduardopelitti) [inline autolink/postlink/activityLinker.js:59]: For some cases, the script was running silently and failing/not doing anything without proper feedback. I added some messages for this as well.

File: autolink/postlink/activityLinker.js [hunk 6]
```diff
@@ -27,24 +56,32 @@ class ActivityLinker {
...
-      warnn("   MainActivity already extends NavigationActivity")
+        .replace("import com.facebook.react.ReactActivity;", "import com.reactnativenavigation.NavigationActivity;");
     }
-    return activityContent;
+    if (this._hasAlreadyExtendReactActivity) {
+      warnn("   MainActivity already extends NavigationActivity");
+      return activityContent;
+    }
+
+    throw new Error("MainActivity was not successfully replaced. Please check the documentation and proceed manually.");
```
Comment (eduardopelitti) [inline autolink/postlink/activityLinker.js:75]: This Error will be catched when the individual method step is called from the main link() method in the ActivityLinker Class.

File: autolink/postlink/activityLinker.js (hunk 1 above)
Comment (gaguirre) [reply autolink/postlink/activityLinker.js]: I saw similar variables in other files with `Success` as a suffix.
Should we rename these ones?

File: autolink/postlink/applicationLinker.js [hunk 7]
```diff
@@ -1,68 +1,126 @@
...
 
   _extendNavigationHost(applicationContent) {
     if (this._doesExtendReactNativeHost(applicationContent)) {
-      debugn("   Changing host implementation to NavigationReactNativeHost");
+      debugn('   Changing host implementation to NavigationReactNativeHost')
+      return applicationContent
+        .replace('new ReactNativeHost(this)', 'new NavigationReactNativeHost(this)')
+        .replace(
+          'import com.facebook.react.ReactNativeHost;',
+          'import com.facebook.react.ReactNativeHost;\nimport com.reactnativenavigation.react.NavigationReactNativeHost;'
```
Comment (gaguirre) [inline autolink/postlink/applicationLinker.js]: You're keeping `import com.facebook.react.ReactNativeHost;`
Is that import needed?

File: autolink/postlink/gradleLinker.js [hunk 8]
```diff
@@ -1,60 +1,98 @@
...
+var DEFAULT_KOTLIN_VERSION = '1.3.61'
 // This should be the minSdkVersion required for RNN.
 var DEFAULT_MIN_SDK_VERSION = 19
 
 class GradleLinker {
   constructor() {
-    this.gradlePath = path.rootGradle;
+    this.gradlePath = path.rootGradle
+    this.setKlotinVersionSuccess = false
+    this.setKotlinPluginDependency = false
```
Comment (gaguirre) [inline autolink/postlink/gradleLinker.js]: Same here: some variables has `Success` as a suffix, some others don't.
Not sure if it's intentional.

**Review by gaguirre — COMMENTED** [review]
It looks great overall. Great job @eduardopelitti!
I was thinking that having tests would be great. Maybe it's easier than one could expect.

File: autolink/postlink/activityLinker.js (hunk 1 above)
Comment (eduardopelitti) [reply autolink/postlink/activityLinker.js]: Yes, @ignacioola commented the same, I used one criteria first and then ditched it on the rest 😓

File: autolink/postlink/applicationLinker.js (hunk 7 above)
Comment (eduardopelitti) [reply autolink/postlink/applicationLinker.js]: I will check, I tried to leave most of the logic unchanged, aside from things that I saw were clearly wrong or were unused code.

File: autolink/postlink/gradleLinker.js (hunk 8 above)
Comment (eduardopelitti) [reply autolink/postlink/gradleLinker.js]: It's not, I will use a standard naming, appending Success to all.

File: autolink/postlink/applicationLinker.js (hunk 7 above)
Comment (eduardopelitti) [reply autolink/postlink/applicationLinker.js]: It is indeed needed, but Android Studio points out that the import of `import android.app.Application;` and `import com.facebook.soloader.SoLoader;` are not necessary after the script runs, perhaps we could add this in a separate PR?

**Comment by eduardopelitti** [comment]
> It looks great overall. Great job @eduardopelitti!
> I was thinking that having tests would be great. Maybe it's easier than one could expect.

Thanks! Having tests sounds good too. I'll look into it

**Review by ignacioola — APPROVED** [review]

**Review by gaguirre — APPROVED** [review]

**Comment by eduardopelitti** [comment]
Closing this here.

This was merged to Wix's repository so there's no need to merge this here too.

**Pull request closed without merging**